src/partitioning/partitioner.C

```diff
@@ -405,6 +405,306 @@ void Partitioner::set_parent_processor_ids(MeshBase & mesh)
+
+          auto & map_set = processor_pair_to_nodes[std::make_pair(std::min(elem->processor_id(), neigh->processor_id()), std::max(elem->processor_id(), neigh->processor_id()))];
+          for (auto global_node_id: common_nodes)
+            map_set.insert(global_node_id);
+        }
+      }
+
+    }
+
+   std::vector<std::vector<const Elem *>> nodes_to_elem_map;
+
+   MeshTools::build_nodes_to_elem_map(mesh, nodes_to_elem_map);
+
+   std::vector<const Node *>  neighbors;
+   std::set<dof_id_type> neighbors_order;
+   std::vector<dof_id_type> common_nodes;
+   std::queue<dof_id_type> nodes_queue;
+
+   std::vector<PetscInt> rows;
+   std::vector<PetscInt> cols;
+
+   std::map<dof_id_type, dof_id_type> global_to_local;
+
+   for (auto & pmap: processor_pair_to_nodes)
+    {
+      unsigned int i = 0;
+
+      rows.clear();
+      rows.resize(pmap.second.size()+1);
+      cols.clear();
+      for (auto it = pmap.second.begin(); it != pmap.second.end(); it++)
+        global_to_local[*it] = i++;
+
+      i = 0;
+      for (auto it = pmap.second.begin(); it != pmap.second.end(); it++, i++)
+      {
+        auto & node = mesh.node_ref(*it);
+        neighbors.clear();
+        MeshTools::find_nodal_neighbors(mesh, node, nodes_to_elem_map, neighbors);
+        neighbors_order.clear();
+        for (auto & neighbor: neighbors)
+          neighbors_order.insert(neighbor->id());
+
+        common_nodes.clear();
+        std::set_intersection(pmap.second.begin(), pmap.second.end(),
+                         neighbors_order.begin(), neighbors_order.end(),
+                              std::back_inserter(common_nodes));
+
+       rows[i+1] = rows[i] +  common_nodes.size();
+
+       for (auto node : common_nodes)
+         cols.push_back(global_to_local[node]);
+      }
+#if LIBMESH_HAVE_PETSC
+    Mat adj;
+    MatPartitioning part;
+    IS is;
+    PetscInt local_size;
+    PetscInt *adj_i, *adj_j;
+    const PetscInt *indices;
+    PetscCalloc1(rows.size(), &adj_i);
+    PetscCalloc1(cols.size(), &adj_j);
+    for (PetscInt ii=0; ii<rows.size(); ii++)
+      adj_i[ii] = rows[ii];
+
+    for (PetscInt ii=0; ii<rows.size(); ii++)
+      adj_j[ii] = cols[ii];
+
+    MatCreateMPIAdj(PETSC_COMM_SELF, pmap.second.size(), pmap.second.size(), adj_i, adj_j,nullptr,&adj);
+    MatPartitioningCreate(PETSC_COMM_SELF,&part);
+    MatPartitioningSetAdjacency(part,adj);
+    MatPartitioningSetType(part,"ptscotch");
+    MatPartitioningSetFromOptions(part);
+    MatPartitioningApply(part,&is);
+
+    MatDestroy(&adj);
+    MatPartitioningDestroy(&part);
+
+    ISGetLocalSize(is, &local_size);
+    ISGetIndices(is, &indices);
+    i = 0;
+    for (auto it = pmap.second.begin(); it != pmap.second.end(); it++, i++)
+    {
+      auto & node = mesh.node_ref(*it);
+      if (indices[i])
+        node.processor_id() = pmap.first.second;
+      else
+        node.processor_id() = pmap.first.first;
+    }
+    ISRestoreIndices(is, &indices);
+    ISDestroy(&is);
+#endif
```

Comment:
This entire function should be wrapped in `#ifdef LIBMESH_HAVE_PETSC` and should throw a `libmesh_error_msg()` in the `#else` branch.